affordance_primitives/src/task_monitor/task_monitor.cpp

```diff
@@ -0,0 +1,163 @@
+#include <affordance_primitives/task_monitor/task_monitor.hpp>
+
+namespace affordance_primitives
+{
+TaskMonitor::TaskMonitor(const ros::NodeHandle& nh, const std::string ft_topic_name) : nh_(nh), result_(STOP_REQUESTED)
+{
+  ft_sub_ = nh_.subscribe(ft_topic_name, 1, &TaskMonitor::ftCB, this);
+};
+
+TaskMonitor::~TaskMonitor()
+{
+  stopMonitor();
+}
+
+void TaskMonitor::startMonitor(const AffordanceParameter& parameters, const double timeout)
+{
+  {
+    // Save variables for starting the move
+    const std::lock_guard<std::mutex> lock(mutex_);
+    parameters_ = parameters;
+    continue_monitoring_ = true;
+
+    // Reset the result
+    result_ = INVALID_RESULT;
+
+    // Set end time if passed a timeout parameter
+    if (timeout > 0)
+    {
+      has_timeout_ = true;
+      end_time_ = ros::Time::now() + ros::Duration(timeout);
+    }
+    else
+    {
+      has_timeout_ = false;
+    }
+  }
+
+  if (!active_monitor_)
+  {
+    // Start the monitoring
+    if (thread_.joinable())
+    {
+      thread_.join();
+    }
+    thread_ = std::thread([this] { mainLoop(); });
+  }
+
+  // Wait for monitor to go active
+  std::unique_lock<std::mutex> lk(mutex_);
+  condition_variable_.wait(lk, [this] { return active_monitor_ == true; });
+}
+
+ExecutionResult TaskMonitor::waitForResult()
+{
+  // All we really have to do is wait for the thread to finish
+  if (thread_.joinable())
+  {
+    thread_.join();
+  }
+
+  return result_;
+}
+
+std::optional<ExecutionResult> TaskMonitor::getResult()
+{
+  // If the monitor is still running, get outta here
+  if (active_monitor_)
+  {
+    return std::nullopt;
+  }
+
+  // Otherwise, just get the ready result
+  return result_;
+}
+
+void TaskMonitor::stopMonitor()
+{
+  // Request a stop and wait for it to happen
+  continue_monitoring_ = false;
+  if (thread_.joinable())
+  {
+    thread_.join();
+  }
+}
+
+void TaskMonitor::ftCB(const WrenchStamped& msg)
+{
+  last_wrench_msgs_.push(msg);
+}
+
+std::optional<ExecutionResult> TaskMonitor::checkForViolaton(const AffordanceParameter& params)
+{
+  while (!last_wrench_msgs_.empty())
+  {
+    auto wrench = last_wrench_msgs_.front().wrench;
```

Comment:
`auto` is great, however I tend to not use it since it could decrease readability by other programmers (just wanted to mention that here)

Reply:
I tend to agree, but here I will leave it because the `.wrench` is pretty clear, and I want to reduce number of times I encode message types (because all of those places need to change for ROS 2 port)